core-primitives/nonce-cache/src/nonce_cache.rs

```diff
@@ -0,0 +1,101 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+use std::sync::SgxRwLock as RwLock;
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+use std::sync::SgxRwLockWriteGuard as RwLockWriteGuard;
+
+#[cfg(feature = "std")]
+use std::sync::RwLock;
+#[cfg(feature = "std")]
+use std::sync::RwLockWriteGuard;
+
+use crate::{
+	error::{Error, Result},
+	GetNonce, MutateNonce, Nonce,
+};
+
+/// Local nonce cache
+///
+/// stores the nonce internally, protected by a RW lock for concurrent access
+#[derive(Default)]
+pub struct NonceCache {
+	nonce_lock: RwLock<Nonce>,
+}
+
+impl NonceCache {
+	pub fn new(nonce_lock: RwLock<Nonce>) -> Self {
+		NonceCache { nonce_lock }
+	}
+}
+
+impl MutateNonce for NonceCache {
+	fn load_for_mutation(&self) -> Result<RwLockWriteGuard<'_, Nonce>> {
+		self.nonce_lock.write().map_err(|_| Error::LockPoisoning)
+	}
+}
+
+impl GetNonce for NonceCache {
+	fn get_nonce(&self) -> Result<Nonce> {
+		let nonce_lock = self.nonce_lock.read().map_err(|_| Error::LockPoisoning)?;
+		Ok(*nonce_lock)
+	}
+}
```

Comment:
The nonce cache is basically a wrapper around a `RwLock`, with a bit of convenience added to it.